File: frontend/src/types/device.ts
export type DeviceStatus = "ON" | "OFF" | "UNKNOWN";

export interface Device {
  id: number;
  device_id: string;
  device_name: string;
  room: string;
  status: DeviceStatus;
  online: boolean;
  ip_address: string | null;
  mac_address: string | null;
  firmware: string | null;
  rssi: number | null;
  last_seen: string | null;
  created_at: string;
}

export interface PaginatedResponse<T> {
  items: T[];
  total: number;
  page: number;
  page_size: number;
}

export interface DeviceUpdateMessage {
  type: "device_update";
  data: Device;
}

File: backend/app/models/device.py
import enum
from datetime import datetime

from sqlalchemy import Boolean, Enum, Integer, String
from sqlalchemy.orm import Mapped, mapped_column
from sqlalchemy.sql import func

from app.database.base import Base


class DeviceStatus(str, enum.Enum):
    ON = "ON"
    OFF = "OFF"
    UNKNOWN = "UNKNOWN"


class Device(Base):
    __tablename__ = "devices"

    id: Mapped[int] = mapped_column(primary_key=True)
    device_id: Mapped[str] = mapped_column(String(64), unique=True, index=True)
    device_name: Mapped[str] = mapped_column(String(128))
    room: Mapped[str] = mapped_column(String(64))
    status: Mapped[DeviceStatus] = mapped_column(
        Enum(DeviceStatus, name="device_status_enum"), default=DeviceStatus.UNKNOWN
    )
    online: Mapped[bool] = mapped_column(Boolean, default=False, index=True)
    ip_address: Mapped[str | None] = mapped_column(String(45), nullable=True)
    mac_address: Mapped[str | None] = mapped_column(String(17), nullable=True)
    firmware: Mapped[str | None] = mapped_column(String(32), nullable=True)
    rssi: Mapped[int | None] = mapped_column(Integer, nullable=True)
    last_seen: Mapped[datetime | None] = mapped_column(nullable=True)
    created_at: Mapped[datetime] = mapped_column(server_default=func.now())
